packages/block-editor/src/hooks/block-style-variation.js

```diff
@@ -0,0 +1,136 @@
+	// editor settings and available in the post editor.
+	// This can be updated once the global styles data is consistently
+	// available across the editors.
+	const { merged: mergedConfig } = useContext( GlobalStylesContext );
+	const { globalSettings, globalStyles } = useSelect( ( select ) => {
+		const { __experimentalFeatures, __experimentalStyles } =
+			select( blockEditorStore ).getSettings();
+		return {
+			globalSettings: __experimentalFeatures,
+			globalStyles: __experimentalStyles,
+		};
+	}, [] );
+
+	return useMemo( () => {
+		const styles = mergedConfig?.styles ?? globalStyles;
+		const variationStyles =
+			styles?.blocks?.[ name ]?.variations?.[ variation ];
+
+		return {
+			settings: mergedConfig?.settings ?? globalSettings,
+			// The variation style data is all that is needed to generate
+			// the styles for the current application to a block. The variation
+			// name is updated to match the instance specific class name.
+			styles: {
+				blocks: {
+					[ name ]: {
+						variations: {
+							[ `${ variation }-${ clientId }` ]: variationStyles,
+						},
+					},
+				},
+			},
+		};
+	}, [
+		mergedConfig,
+		globalSettings,
+		globalStyles,
+		variation,
+		clientId,
+		name,
+	] );
+}
+
+// Rather than leveraging `useInstanceId` here, the `clientId` is used.
+// This is so that the variation style override's ID is predictable
+// when the order of applied style variations changes.
+function useBlockProps( { name, className, clientId } ) {
+	const variation = getVariationNameFromClass( className );
+	const variationClass = `is-style-${ variation }-${ clientId }`;
+
+	const { getBlockStyles } = useSelect( blocksStore );
+
+	const { settings, styles } = useBlockSyleVariation(
+		name,
+		variation,
+		clientId
+	);
+
+	const variationStyles = useMemo( () => {
+		if ( ! variation ) {
+			return;
+		}
+
+		const variationConfig = { settings, styles };
+		const blockSelectors = getBlockSelectors(
+			getBlockTypes(),
+			getBlockStyles,
+			clientId
+		);
+		const hasBlockGapSupport = false;
+		const hasFallbackGapSupport = true;
+		const disableLayoutStyles = true;
+		const isTemplate = true;
+
+		return toStyles(
+			variationConfig,
+			blockSelectors,
+			hasBlockGapSupport,
+			hasFallbackGapSupport,
+			disableLayoutStyles,
+			isTemplate,
+			{
+				blockGap: false,
+				blockStyles: true,
+				layoutStyles: false,
+				marginReset: false,
+				presets: false,
+				rootPadding: false,
+			}
+		);
+	}, [ variation, settings, styles, getBlockStyles, clientId ] );
+
+	useStyleOverride( {
+		id: `variation-${ clientId }`,
+		css: variationStyles,
+		__unstableType: 'variation',
+		// The clientId will be stored with the override and used to ensure
+		// the order of overrides matches the order of blocks so that the
+		// correct CSS cascade is maintained.
+		clientId,
+	} );
+
+	return variation ? { className: variationClass } : {};
+}
+
+export default {
+	hasSupport: () => true,
```

Comment:
Are variations supported by all blocks?

Reply:
That was my understanding as any block type could have a block style registered for it.

I do believe it makes more sense for some blocks than others though.